src/collective/listmonk/services/subscriptions.py

```diff
@@ -0,0 +1,216 @@
+from BTrees.OOBTree import OOBTree
+from datetime import datetime
+from datetime import timezone
+from email.message import EmailMessage
+from email.utils import formataddr
+from plone import api
+from plone.base.interfaces.controlpanel import IMailSchema
+from plone.registry.interfaces import IRegistry
+from plone.restapi.services import Service
+from plone.restapi.testing import RelativeSession
+from typing import Optional
+from typing import TypeVar
+from urllib.parse import quote
+from zExceptions import BadRequest
+from zope.component import getUtility
+
+import json
+import pydantic
+import uuid
+
+
+class SubscriptionRequest(pydantic.BaseModel):
+    list_ids: list[int]
+    name: str
+    email: str
+
+
+class PendingConfirmation(pydantic.BaseModel):
+    token: str
+    list_ids: list[int]
+    sub_id: int
+    created_at: datetime = pydantic.Field(
+        default_factory=lambda: datetime.now(timezone.utc)
+    )
+
+
+T = TypeVar("T", bound=pydantic.BaseModel)
+
+
+class PydanticService(Service):
+    def validate(self, model: type[T]) -> T:
```

Comment:
Loved this approach

Reply:
I wanted to make the input model an attribute of the service class, but I couldn't figure out how to make the generic type hints work correctly that way.